flytekit/extras/tasks/shell.py

```diff
@@ -192,26 +199,54 @@ def execute(self, **kwargs) -> typing.Any:
         Executes the given script by substituting the inputs and outputs and extracts the outputs from the filesystem
         """
         logger.info(f"Running shell script as type {self.task_type}")
+        breakpoint()
+
+        # somewhat hidden behavior and hacky at the moment, if script_args is in kwargs, grab it and save it
+        script_args = kwargs.pop("script_args") if "script_args" in kwargs else None
+
         if self.script_file:
             with open(self.script_file) as f:
                 self._script = f.read()
 
+        # In the event ENV variables are used in the script AND as part of the outputs, need to process them first
         outputs: typing.Dict[str, str] = {}
         if self._output_locs:
             for v in self._output_locs:
-                outputs[v.var] = self._interpolizer.interpolate(v.location, inputs=kwargs)
+                outputs[v.var] = self.__output_interpolizer.interpolate(v.location, inputs=kwargs)
+
+        # remove the extra ENV vars from kwargs and set them for the shell script
+        original_env = os.environ.copy()
+        if self._env:
+            for k in self._env:
+                if k in kwargs:
```

Comment:
Oh I see, you as passing them in as seperate kwargs. I was thinking you would just pass a dict in as one of the kwargs for execute (e.g., env = {} gets passed when calling the callable). That explains the typing above

Reply:
Not sure which is better... on one hand, if you pass a dict you need to either use a dynamic workflow to assemble the dict, or use a task to assemble it (i.e., you can't have a dict of flyte objects I think, the entire dict needs to be a flyte object/promise), however, this is probably harder to read because the env variables are mixed into the kwargs of the callable (so unless you prefix them all like "ENV_WHATEVER" so you can distinguish them, its harder to read)

Reply:
My thinking was "Convention: All environment variables are UPPERCASE"

Reply:
passing a dict to the `inputs` could probably work but we can't restrict the set of keys at compile time, so any env vars could be added and there wouldn't be control over it. I'm not sure how complex inputs are handled on the flyte console UI when specifying params at launch